### PR TITLE
refactor(stdlib): add null support to join function

### LINKED ISSUE #668: Update join to understand Null

### PR DESCRIPTION
**Pull request by wolffcm**
Fixes #668.

### Done checklist
- [ ] docs/SPEC.md updated
- [ ] Test cases written

### CODE REVIEW

File: stdlib/universe/join.go [hunk 1]
```diff
@@ -808,6 +828,23 @@ func (c *MergeJoinCache) buildPostJoinSchema() {
...
+// equalJoinKeys compares two keys for equality.
+// Null values are not considered equal when joining (unlike when grouping).
+func equalJoinkeys(left, right flux.GroupKey) bool {
+	for _, v := range left.Values() {
+		if v.IsNull() {
+			return false
+		}
+	}
+	for _, v := range right.Values() {
+		if v.IsNull() {
```
Comment (jlapacik) [inline stdlib/universe/join.go]: Is this necessary after you added the `isNull()` check to `values.Equal()`? That is, can you reduce this function to just `left.Equal(right)`?

File: stdlib/universe/join.go [hunk 2]
```diff
@@ -697,7 +699,25 @@ func (c *MergeJoinCache) insertIntoBuffer(id execute.DatasetID, tbl flux.Table)
...
+
+	// Optimization: if any group key columns overlap join key columns,
+	// and there are any nulls in those columns, we can discard this table,
+	// since null != null for joining purposes.
+	k := tbl.Key()
+	for j, col := range k.Cols() {
+		if c.on[col.Label] {
+			if k.IsNull(j) {
+				// Discard the table and return.
+				err := tbl.Do(func(flux.ColReader) error {
```
Comment (jlapacik) [inline stdlib/universe/join.go]: Why is the explicit call to `tbl.Do` necessary? Can we not just return right here, not caching the table?

File: stdlib/universe/join.go (hunk 1 above)
Comment (wolffcm) [reply stdlib/universe/join.go]: Yes, you're right.  I'll fix it.

File: stdlib/universe/join.go (hunk 2 above)
Comment (wolffcm) [reply stdlib/universe/join.go]: Apparently we need to iterate over the table for it to be consumed.  Lorenzo filed an issue for this:
https://github.com/influxdata/flux/issues/643
I'm not quite clear on why this is exactly, though.  There are a few different implementions of `flux.Table`, I think some of them may require their contents to be iterated exactly once for the pipeline to work right.

I'll add a comment to the code.

**Pull request merged**